How to/manage healthcare element (#12)
icure/medtech-docs: sdks/how-to/how-to-manage-healthcare-elements.md

)
```

If the operation is successful, the method returns the updated Healthcare Element as it is stored in the Backend.

:::caution

To update a Healthcare Element, both id and rev fields must be valid.

:::

## Retrieving Healthcare Element Using Complex Search Criteria

If you want to retrieve a set of Healthcare Element that satisfy complex criteria, you can use a Filter.  
In the following example, you will instantiate a filter to retrieve all the Healthcare Element that a Healthcare Professional
 can access

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:create HE filter-->
```typescript
const healthcareElementFilter = await new HealthcareElementFilter()
  .forDataOwner(user.healthcarePartyId)
  .forPatients(api.cryptoApi, [patient])
  .build()
```

:::note

You can learn more about filters in the how to.

:::

After creating a filter, you can use it to retrieve the Healthcare Elements.

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:use HE filter method-->
```typescript
const healthcareElementsFirstPage = await api.healthcareElementApi.filterHealthcareElement(
  healthcareElementFilter,
  undefined,
  10
)
```

> Note that the Filter method will return a PaginatedList which contains a number of element up to the maximum you specify

Reviewer: ```suggestion
The `filter` method returns a PaginatedList that contains at most the number of elements stated
```

 in the method's parameter. If you do not specify any number, the default value is 1000.  
To retrieve more Healthcare Elements, you can call the same method again, using the startDocumentId provided in the previous
 request.

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:use HE filter method second page-->
```typescript
const healthcareElementsSecondPage = await api.healthcareElementApi.filterHealthcareElement(
  healthcareElementFilter,
  healthcareElementsFirstPage.nextKeyPair.startKeyDocId,
  10
)
```

If the `nextKeyPair` property of the result is `undefined`, than there are no more Healthcare Elements to retrieve.  
You can also retrieve just the id of the Healthcare Element instead of the whole documents by using the match method.

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:use HE match method-->
```typescript
const healthcareElementsIdList = await api.healthcareElementApi.matchHealthcareElement(healthcareElementFilter)
```

> Finally, you can also retrieve all the Healthcare Elements related to a specific patient that the current Data Owner 

Reviewer: ```suggestion
You can also retrieve all the Healthcare Elements belonging to a specific patient that the current Data Owner 
```

  new HealthcareElement({
    description: 'The patient recovered',
    openingDate: new Date('2020-11-08').getTime(),
    healthcareElementId: startHealthcareElement.healthcareElementId,
  }),
  patient.id,
)
```

:::note

The `healthcareElementId` is the id of the first Healthcare Element of the series.

:::

Several unrelated Healthcare Elements can also be created at once.

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:create multiple HEs as data owner-->
```typescript
const healthcareElement1 = new HealthcareElement({
  description: 'The patient has been diagnosed Pararibulitis',
  codes: new Set([
    new CodingReference({
      id: 'SNOMEDCT|617|20020131',
      type: 'SNOMEDCT',
      code: '617',
      version: '20020131',
    }),
  ]),
  openingDate: new Date('2019-10-12').getTime(),
})

const healthcareElement2 = new HealthcareElement({
  description: 'The patient has also the flu',
  openingDate: new Date('2020-11-08').getTime(),
})

const newElements = await api.healthcareElementApi.createOrModifyHealthcareElements(
  [healthcareElement1, healthcareElement2],
  patient.id,
)
```

:::caution

> Even if you associate a Healthcare Element to a Patient, the Patient does not have access to it unless you give access 

Reviewer 2: Even if you associate a Healthcare Element to a Patient, the Patient does not automatically have access to it. You need to explicitly give access to the patient user to this created Healthcare Element by calling the service `giveAccessTo`.

to them.

:::

## Sharing a Healthcare Element with a Patient

> After creating the Healthcare Element, the Healthcare Professional shares it with the Patient.

Reviewer 2: After creating the Healthcare Element, the Healthcare Professional can share it with the Patient.

Plus, add more detail / a note to explain that a Patient can give access to a HCP, to another patient, a hcp can give access to a patient, to another hcp, ... it's not one way

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:HE sharing with data owner-->
```typescript
const sharedHealthcareElement = await api.healthcareElementApi.giveAccessTo(
  healthcareElement,
  patient.id,
)
```

If the operation is successful, the method returns a Promise with the updated Healthcare Element.

## Retrieving a Healthcare Element Using its ID

> A single Healthcare Element can be retrieved from the Backend using its id.

Reviewer 2: A single Healthcare Element can be retrieved from the iCure Back-end using its id.

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:retrieve a HE as data owner-->
```typescript
const healthcareElementToRetrieve = new HealthcareElement({
  description: 'To retrieve, I must create',
})

> const createdHealthcareElement = await api.healthcareElementApi.createOrModifyHealthcareElement(

Reviewer 2: I would not put the create in the code sample : Reuse the one you previously created in the other code samples

  healthcareElementToRetrieve,
  patient.id,
)

const retrievedHealthcareElement = await api.healthcareElementApi.getHealthcareElement(
  createdHealthcareElement.id,
)
```

:::caution

Trying to retrieve a Healthcare Element you do not have access to will produce an error.

:::

## Modifying a Healthcare Element

Given an existing Healthcare Element, it is possible to modify it.  

> Some fields, like id and rev, cannot be modified.

Reviewer 2: Precise all fields that can't be modified.

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:modify a HE as data owner-->
```typescript
const yetAnotherHealthcareElement = await api.healthcareElementApi.createOrModifyHealthcareElement(
  new HealthcareElement({
    description: 'To modify, I must create',
  }),
  patient.id,
)

const modifiedHealthcareElement = {
  ...yetAnotherHealthcareElement,
  description: 'I can change and I can add',
  openingDate: new Date('2019-10-12').getTime(),
}

const modificationResult = await api.healthcareElementApi.createOrModifyHealthcareElement(
  modifiedHealthcareElement,
  patient.id,
)
```

> If the operation is successful, the method returns the updated Healthcare Element as it is stored in the Backend.

Reviewer 2: If the operation is successful, the method returns the updated Healthcare Element.

:::caution

> To update a Healthcare Element, both id and rev fields must be valid.

Reviewer 2: What does it mean that id and rev must be valid ? Explain to the user

:::

## Retrieving Healthcare Element Using Complex Search Criteria

If you want to retrieve a set of Healthcare Element that satisfy complex criteria, you can use a Filter.  
In the following example, you will instantiate a filter to retrieve all the Healthcare Element of a Patient that a Healthcare Professional
 can access

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:create HE filter-->
```typescript
const healthcareElementFilter = await new HealthcareElementFilter()
  .forDataOwner(user.healthcarePartyId)
  .forPatients(api.cryptoApi, [patient])
  .build()
```

:::note

> You can learn more about filters in the how to.

Reviewer 2: I think there's already the empty file, so that you can add the link, even if it is empty for now

:::

After creating a filter, you can use it to retrieve the Healthcare Elements.

<!-- file://code-samples/how-to/manage-healthcare-elements/index.mts snippet:use HE filter method-->
```typescript
const healthcareElementsFirstPage = await api.healthcareElementApi.filterHealthcareElement(
  healthcareElementFilter,
  undefined,
  10,
)
```

The `filter` method returns a PaginatedList that contains at most the number of elements stated
 in the method's parameter. If you do not specify any number, the default value is 1000.  

> To retrieve more Healthcare Elements, you can call the same method again, using the startDocumentId provided in the previous

Reviewer 2: To retrieve more Healthcare Elements, you can call the same method again, using the startDocumentId provided in the previous PaginatedList.
